File: leetcode-submissions/675_Cut_Off_Trees_for_Golf_Event.cpp
class Solution {
public:
    int m , n ;
    int bfs(int dx , int dy , int tx , int ty, vector<vector<int>>&forest){
        if (dx == tx && dy == ty) return 0;
        vector<vector<bool>>visited(m , vector<bool>(n , false));
        queue<tuple< int ,int, int >>q;
        q.push({dx , dy, 0 });
        visited[dx][dy]= true ;
        vector<vector<int>>directions = {{0 , 1}, { 1, 0 }, {0 ,-1}, {-1 , 0 }};
        
        while(not q.empty()){
            auto [ i , j , d]= q.front();
            q.pop();
            for(auto& dir : directions ){
                int i_ = i +dir[0];
                int j_ = j + dir[1];
                if(i_<m and j_<n and i_>=0 and j_>=0 and visited[i_][j_]==false and forest[i_][j_]!=0){
                    if(i_==tx and j_==ty){
return d+1; }
                    visited[i_][j_]=true ;
                    q.push({i_, j_, d+1});
                    
                }
            }
            
        }
        return -1;
    }
    int cutOffTree(vector<vector<int>>& forest) {
        vector<vector<int>>trees;
         m = forest.size(), n = forest[0].size();
        for(int i =0 ;i<m;i++){
            for(int j = 0 ;j<n;j++){
               if(forest[i][j]>1) trees.push_back({forest[i][j], i , j});
            }
        }
        sort(trees.begin(), trees.end());
        int dx = 0 , dy =0 , ans=0 ;
        for(auto tree : trees){
            int tx = tree[1], ty = tree[2];
            
             int d  = bfs(dx , dy ,tx , ty, forest );
            if(d==-1){
                return -1;
            }
            ans+=d;
            dx = tx;
            dy = ty;
        }
       
        return ans ;
    }
};

File: leetcode-submissions/675_Cut_Off_Trees_for_Golf_Event.ts
function cutOffTree(forest: number[][]): number {
    if(forest[0][0]===0) return -1; 
    const m = forest.length, n = forest[0].length;
    let distance  = -1;
    const directions:[number , number][] = [[0 , 1] , [1, 0 ], [0 , -1], [ -1 , 0 ]];
    const queue:[number , number   , number][]  =[[0 , 0, 0 ]];
    const trees :[number , number , number][]  = [];
    for(let i:number =0 ;i<m;i++){
        for(let j:number =0;j<n;j++){
            if(forest[i][j]>1){
                trees.push([forest[i][j], i , j]);
            }
            
        }
    }
    trees.sort((a, b)=>a[0]-b[0]);
    function bfs(sx: number, sy: number, tx: number, ty: number): number {
        if (sx === tx && sy === ty) return 0;
        const visited = Array.from({ length: m }, () => Array(n).fill(false));
        const queue: [number, number, number][] = [[sx, sy, 0]];
        visited[sx][sy] = true;
        const dirs = [[0,1], [1,0], [0,-1], [-1,0]];

        while (queue.length) {
            const [x, y, d] = queue.shift()!;
            for (const [dx, dy] of dirs) {
                const nx = x + dx, ny = y + dy;
                if (
                    nx >= 0 && ny >= 0 && nx < m && ny < n &&
                    !visited[nx][ny] && forest[nx][ny] !== 0
                ) {
                    if (nx === tx && ny === ty) return d + 1;
                    visited[nx][ny] = true;
                    queue.push([nx, ny, d + 1]);
                }
            }
        }
        return -1; // target unreachable
    }
        let totalDistance =0, currx=0 , curry=0  ;
    for(const [_ , i , j] of trees){
        let d = bfs(currx, curry , i , j);
        if(d===-1) return -1;
        
        totalDistance +=d;
        currx = i ;
        curry = j ;
    }
    return totalDistance ;
        
        
    
};
